src/tasks/env.rs

```diff
@@ -0,0 +1,115 @@
+    provider: RuProvider,
+}
+
+impl EnvTask {
+    /// Create a new EnvTask with the given config and provider.
+    pub const fn new(config: BuilderConfig, provider: RuProvider) -> Self {
+        Self { config, provider }
+    }
+
+    /// Construct a BlockEnv by making calls to the provider.
+    pub fn construct_block_env(&self, previous: &Header) -> BlockEnv {
+        BlockEnv {
+            number: previous.number + 1,
+            beneficiary: self.config.builder_rewards_address,
+            // NB: EXACTLY the same as the previous block
+            timestamp: previous.number + self.config.slot_calculator.slot_duration(),
+            gas_limit: self.config.rollup_block_gas_limit,
+            basefee: previous
+                .next_block_base_fee(BaseFeeParams::ethereum())
+                .expect("signet has no non-1559 headers"),
+            difficulty: U256::ZERO,
+            prevrandao: Some(B256::random()),
+            blob_excess_gas_and_price: Some(BlobExcessGasAndPrice {
+                excess_blob_gas: 0,
+                blob_gasprice: 0,
+            }),
+        }
+    }
+
+    /// Construct the BlockEnv and send it to the sender.
+    pub async fn task_fut(self, sender: watch::Sender<Option<BlockEnv>>) {
+        let span = info_span!("EnvTask::task_fut::init");
+        let mut poller = match self.provider.watch_blocks().instrument(span.clone()).await {
+            Ok(poller) => poller,
+            Err(err) => {
+                let _span = span.enter();
+                error!(%err, "Failed to watch blocks");
+                return;
+            }
+        };
+
+        poller.set_poll_interval(Duration::from_millis(250));
+
+        let mut blocks = poller.into_stream();
+
+        while let Some(blocks) =
+            blocks.next().instrument(info_span!("EnvTask::task_fut::stream")).await
+        {
+            let Some(block) = blocks.last() else {
+                // Clear the env, as the block has changed but we can't
+                // get a new one
+                debug!("empty filter changes");
+                let _ = sender.send(None);
+                continue;
+            };
+            let span = info_span!("EnvTask::task_fut::loop", hash = %block, number = tracing::field::Empty);
+
+            let previous = match self
+                .provider
+                .get_block((*block).into())
+                .into_future()
+                .instrument(span.clone())
+                .await
+            {
+                Ok(Some(block)) => block.header.inner,
+                Ok(None) => {
+                    let _span = span.enter();
+                    let _ = sender.send(None);
+                    debug!("block not found");
+                    continue;
+                }
+                Err(err) => {
+                    let _span = span.enter();
+                    let _ = sender.send(None);
+                    error!(%err, "Failed to get latest block");
+                    break;
```

Comment:
it's fine to break here, since not being able to get blocks means this task would not get any new data and therefore be useless